app/src/studio/gatt_rpc_transport.c

```diff
@@ -0,0 +1,222 @@
+        }
+
+        bt_conn_unref(conn);
+    }
+#endif
+}
+
+static ssize_t read_rpc_resp(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+                             uint16_t len, uint16_t offset) {
+
+    LOG_DBG("Read response for length %d at offset %d", len, offset);
+    return 0;
+}
+
+static ssize_t write_rpc_req(struct bt_conn *conn, const struct bt_gatt_attr *attr, const void *buf,
+                             uint16_t len, uint16_t offset, uint8_t flags) {
+    if (handling_rx) {
+        uint32_t copied = 0;
+        struct ring_buf *rpc_buf = zmk_rpc_get_rx_buf();
+        do {
+            uint8_t *buffer;
+            uint32_t claim_len = ring_buf_put_claim(rpc_buf, &buffer, len - copied);
+
+            if (claim_len > 0) {
+                memcpy(buffer, ((uint8_t *)buf) + copied, claim_len);
+                copied += claim_len;
+            }
+
+            ring_buf_put_finish(rpc_buf, claim_len);
+        } while (len - copied > 0);
+
+        zmk_rpc_rx_notify();
+    }
+
+    return len;
+}
+
+BT_GATT_SERVICE_DEFINE(
+    rpc_interface, BT_GATT_PRIMARY_SERVICE(BT_UUID_DECLARE_128(ZMK_STUDIO_BT_SERVICE_UUID)),
+    BT_GATT_CHARACTERISTIC(BT_UUID_DECLARE_128(ZMK_STUDIO_BT_RPC_CHRC_UUID),
+                           BT_GATT_CHRC_WRITE | BT_GATT_CHRC_READ | BT_GATT_CHRC_INDICATE,
+                           BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT, read_rpc_resp,
+                           write_rpc_req, NULL),
+    BT_GATT_CCC(rpc_ccc_cfg_changed, BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT));
+
+static void refresh_notify_size(void) {
+    struct bt_conn *conn = zmk_ble_active_profile_conn();
+
+    uint8_t ns = 23; // Default MTU size unless negotiated higher
+    if (conn) {
+
+        struct bt_conn_info conn_info;
+        if (bt_conn_get_info(conn, &conn_info) >= 0) {
+            ns = conn_info.le.data_len->tx_max_len;
+        }
+        bt_conn_unref(conn);
+    }
+
+    atomic_set(&notify_size, ns);
+}
+
+static int gatt_start_rx() {
+    refresh_notify_size();
+    handling_rx = true;
+    return 0;
+}
+
+static int gatt_stop_rx(void) {
+    handling_rx = false;
+    return 0;
+}
+
+static struct bt_gatt_indicate_params rpc_indicate_params = {
+    .attr = &rpc_interface.attrs[1],
+};
+
+static void notif_rpc_tx_cb(struct k_work *work) {
+    struct bt_conn *conn = zmk_ble_active_profile_conn();
+    struct ring_buf *tx_buf = zmk_rpc_get_tx_buf();
+
+    if (!conn) {
+        LOG_WRN("No active connection for queued data, dropping");
+        ring_buf_reset(tx_buf);
+        return;
+    }
+
+    uint8_t notify_size = 23; // Default MTU size unless negotiated higher
+    struct bt_conn_info conn_info;
+    if (bt_conn_get_info(conn, &conn_info) >= 0) {
+        notify_size = conn_info.le.data_len->tx_max_len;
+    }
+
+    uint8_t notify_bytes[notify_size];
+
+    while (ring_buf_size_get(tx_buf) > 0) {
+        uint8_t added = 0;
```

Comment:
Any reason to specifically use uint8_t here instead of something like int? (We're right between a couple of probably large variables on the stack, so probably not going to be packing variables together and saving stack size here.)

Reply:
Nope. Changed.